blockstore/apps/bundles/store.py

```diff
@@ -0,0 +1,232 @@
+        For a given django.core.files.File object, return a 20-byte BLAKE2 hash.
+        """
+        hash = hashlib.blake2b(digest_size=20)
+        for chunk in file.chunks():
+            hash.update(chunk)
+        return hash
+
+# @attr.s(frozen=True)
+# class Link():
+#     """
+#     All Link references at this layer should be to BundleSnapshots, not
+#     BundleVersions.
+#     """
+#     alias = attr.ib(type=str)
+#     target = attr.ib()
+#     uses = attr.ib(type=list)
+
+
+# @attr.s()
+# class DependencyMapping():
+#     # Map of strings (no /'s allowed) to BundleVersionKey objects
+#     links = attr.ib(type=dict)
+
+#     def add_link(self, link):
+#         """"""
+#         pass
+
+#     def all(self):
+#         pass
+
+
+@attr.s()
+class BundleSnapshot():
+    """
+    A BundleDataStore doesn't have to store this structure explicitly, as long
+    as it can derive it and return when its summary() method is called.
+    """
+    bundle_uuid = attr.ib()
+
+    # Map of paths to FileInfo objects
+    files = attr.ib(type=dict)
+
+    # 20-byte BLAKE2 hash
+    hash_digest = attr.ib(type=bytes)
+
+    @classmethod
+    def create(cls, bundle_uuid, files):
+        str_to_be_hashed = json.dumps(
+            [bundle_uuid, sorted(files.items())],
+            cls=BundleJSONEncoder,
+            indent=None,
+            separators=(',', ':'),
+        )
+        hash_digest = hashlib.blake2b(str_to_be_hashed.encode('utf-8'), digest_size=20).digest()
+        return cls(bundle_uuid, files, hash_digest)
+
+    def url(self, path):
+        """Return a user-accessible URL to download a path from this Snapshot."""
+        file_info = self.files[path]
+        storage_path = '{}/data/{}'.format(
+            self.bundle_uuid, file_info.hash_digest.hex()
+        )
+        return default_storage.url(storage_path)
+
+@contextmanager
+def files_from_disk(bundle_data_path):
+    """
+    Given a pathlib.Path, return file mapping suitable for BundleSnapshot.
+
+    Returned dictionary will map from str to django.core.files.File objects.
+
+    This will recursively scan subdirectories.
+    """
+    paths_to_files = {
+        path.relative_to(*path.parts[:1]): File(open(path, "rb"))
+        for path in bundle_data_path.rglob('*')
+        if path.is_file()
+    }
+    try:
+        yield paths_to_files
+    finally:
+        for open_file in paths_to_files.values():
+            open_file.close()
+
+
+class BundleJSONEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, FileInfo):
+            return [
+                obj.public,
+                obj.size,
+                obj.hash_digest.hex(),
+            ]
+        elif isinstance(obj, uuid.UUID):
+            return str(obj)
+        elif isinstance(obj, BundleSnapshot):
+            return {
+                'bundle_uuid': obj.bundle_uuid,
+                'hash_digest': obj.hash_digest.hex(),
+                'files': obj.files,
+            }
```

Comment:
A document version would be useful in case we need to make a backwards-incompatible change.

Reply:
Oops. Yeah, and that was even in the design doc... Will add, thank you!